[#437] Friend Request Validation for Different User Groups

tyzhou05 opened the pull request:
## Issue Number:

## Self Check List

- [x] Have you double-checked the base branch into which this PR will be merged?
- [x] Have you rebased from the base branch?

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (formatting, renaming)
- [ ] Merge Develop to Main
- [ ] Other (please describe):

## What does this PR do?
* Added validation checks in `UserFriendRequestCreateSerializer` and `UserFriendRequest` to see if the requester and requestee are using the same version (`current_ver`) before allowing a friend request to be sent. Raises a ValidationError with a 403 Forbidden response if they are not.
  * Changes should ensure that different versions cannot send friend requests to each other through both the default (`/api/user/friend-requests/default/`) and regular (`/api/user/friend-requests/`) friend request endpoints.
* Postman: [api/user/friend-requests](https://diivers.postman.co/workspace/WhoAmI~c2317c3a-2ab1-458e-96b5-01f8ca6b983e/request/6673035-c18e22a2-6961-4519-80a5-4441e0196ef9)
* Postman: [api/user/friend-requests/default](https://diivers.postman.co/workspace/WhoAmI~c2317c3a-2ab1-458e-96b5-01f8ca6b983e/request/6673035-5948bd37-2887-4b72-a7f3-c42e01175ce1?tab=body)

## Preview Image

## Further comments

tyzhou05 committed:
initial version validation implementation

tyzhou05 commented:
@njs03332 Started my initial implementation! I first added some validation to `UserFriendRequestCreateSerializer` to check if users are in the same version, and this should raise a 403 Validation Error if they are not. I also added a check in `BaseUserFriendRequestUpdate` to see if the requestor and requestee are on the same version before the friend request goes through. 

Let me know if this looks good or not, or anywhere else to possibly add validation; I will test these changes in Postman with the friend request API!

njs03332 commented on adoorback/account/serializers.py:
```diff
@@ -375,6 +375,20 @@ def get_requester_detail(self, obj):
+            raise serializers.ValidationError("User not found")
+
+        # check if requester and requestee are in the same experiment version
+        if requester.current_ver != requestee.current_ver:
```
After the serializer raises a validation error like this, the view should catch this and send an appropriate response to the frontend (403 + a message like "Users belong to different groups, so a friend request cannot be sent").  

This way, the frontend can recognize that the error occurred due to users being in different groups and display an appropriate message to the user.

tyzhou05 replied:
Got it, added the change!

tyzhou05 committed:
adding check to views

tyzhou05 commented:
@njs03332 Thanks for the review, added the extra validation and updated the description with what I changed and the Postman tests. Let me know if it looks good to merge!

njs03332 commented on adoorback/account/migrations/0017_merge_20250208_1156.py:
```diff
@@ -0,0 +1,14 @@
+# Generated by Django 4.2.14 on 2025-02-08 19:56
```
I don't think we need this merge file since I made one in the latest main branch! (https://github.com/ssm0318/WhoamI-Today-backend/blob/main/adoorback/account/migrations/0017_merge_20250207_2123.py)

njs03332 reviewed (approved):
Looks good to merge except for the comment I left about the migration file! :)

tyzhou05 commented:
Great, thanks! Will get started on the other assigned issues and plan in to send a review request tomorrow 👍

tyzhou05 committed:
deleting redundant migrations
